Modify LU_backward and lu_solve_backward to use linalg_solve_triangular

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63570
* __->__ #63569

This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D32618014](https://our.internmc.facebook.com/intern/diff/D32618014)

lezcano committed:
Modify LU_backward and lu_solve_backward to use linalg_triangular_solve

This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_triangular_solve"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_triangular_solve"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

nikitaved reviewed (approved):
LGTM! Thanks, Mario!

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

mruberry commented:
@pytorchbot ciflow rerun -l ciflow/all -l ciflow/slow

lezcano committed:
Update on "Modify LU_backward and lu_solve_backward to use linalg_solve_triangular"


This PR also rewrites `lu_solve_backward` from scratch going from
solving 5 systems of equations to just 2.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

anjali411 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2857,19 +2857,15 @@ Tensor linalg_qr_backward(const std::vector<torch::autograd::Variable> &grads, c
-        /*transpose=*/false,
+    // Compute rhs_term @ R^{-H}
+    Tensor grad_A = at::linalg_solve_triangular(
+        R.transpose(-2, -1).conj(),
```
can we just call `R.adjoint()` instead?

also looks like `adjoint`'s documentation doesn't render properly https://pytorch.org/docs/master/generated/torch.Tensor.adjoint.html#torch.Tensor.adjoint

Lezcano replied:
It renders the same as all the other tensor methods (for good or bad...)

And sure, I'll change it into using `mH()` once I go over this function again when moving `lu_solve` to `linalg.`.

anjali411 replied:
ok I meant I don't see docs for `torch.adjoint`. Probably missing an entry in some rst file.

mruberry replied:
Would you file an issue?

Lezcano replied:
oh, I had not realised that clicking on adjoint does not take you anywhere. I'll write a fix.

Lezcano replied:
https://github.com/pytorch/pytorch/pull/68869

anjali411 reviewed (approved):
Thank you! This is a cool change. It'd be nice to benchmark the performance.

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32618014).

mruberry reviewed (approved):
Cool!
